refactor(components,protocol-designer): Componentize StepContainer

SyntaxColoring opened the pull request:
## Overview

There's a "Step Button" component in the Protocol Designer design system, soon to be renamed "Step Container". We've been implementing it with a "local" component in `protocol-designer/`, tied up with things like Redux state, as opposed to a "real" component in `components/`. This is a first step to making it a "real" component in `components/` and unifying it with Figma.

Closes AUTH-1750.

## Changelog

This PR:

* Separates the presentational stuff from the more complex Redux state stuff.
* Moves the presentational stuff into `components/` and makes a Storybook story for it.

No functional change is intended.

Remaining known issues, which I'm not addressing in this PR:

* The overflow menu still appears when `size="iconOnly"`; the designs imply it should go away. (But this would make some actions unreachable, so I'm not sure it's intended in the designs. I've asked for clarification in Figma.)
* We're missing the `focus-visible` styling as shown in the designs. I view this as part of a more general thing where the component isn't usable via keyboard overall. I think fixing this would involve a reworking its DOM structure, like https://stackoverflow.com/q/64522551/497934.
* The designs show the vertical size of the button shrinking when the text is hidden, but it currently doesn't.
* It doesn't have a "subtext" line. (This is part of current work for concurrent module actions, EXEC-1688).
* The designs show an error/warning icon separate from the step icon (so 2 icons total), but we currently have the error/warning icon *replace* the step icon.
* It's still using styled-components instead of CSS Modules.

## Test Plan and Hands on Testing

* [x] Various states look OK in Storybook.
* [x] Drag-and-drop still works.
* [x] Icons are still correct.
* [x] Errors are still shown correctly: steps after the error should be semi-transparent.

## Review requests

* Do all of the remaining known issues seem like things that should ideally be fixed, or were any of those intentional? If they should ideally be fixed, I'm happy to help, but how should those be ticketed? AUTH-427?
* See GitHub comments for code-level things.

## Risk assessment

Medium.

SyntaxColoring committed:
Add a story.

SyntaxColoring committed:
Port protocol-designer to use the new component.

SyntaxColoring committed:
Don't show the overflow menu on alt buttons.

SyntaxColoring committed:
Add a new presentation-only component.

SyntaxColoring commented on protocol-designer/src/pages/Designer/ProtocolSteps/Timeline/StepContainer.tsx:
```diff
@@ -4,26 +4,18 @@ import { useTranslation } from 'react-i18next'
-  OverflowBtn,
   SPACING,
-  StyledText,
+  StepContainer as StepContainerPresentation,
```
As implemented, there's `StepContainer` in `components/`, which is presentational-only, and that's used by `StepContainer` in `protocol-designer/`, which adds stuff like Redux state. Obviously I struggled with naming here, which may be a red flag that I'm doing something structurally wrong.

mjhuff replied:
IMO, the components `StepContainer` is _the_ atomic atomic `StepContainer`, because the PD `StepContainer` is a `StepContainer` plus other functionality - at a cursory glance, it looks like it provides modals and an overflow menu.

All this to say, I think the PD `StepContainer` needs a new name. Up to you if you want to do that in this PR or elsewhere, but I think we want to avoid two components with the same name if possible (even if they're in different namespaces), because this historically leads to problems.

jerader replied:
Agreed with @mjhuff here. We should rename the component in PD since the component in PD is the parent that renders `StepContainer` but it encompasses a lot of other logic.

Well actually, it looks like Design System wants the helix component to be called `StepButton` so maybe leave PD's component as `StepContainer` and rename the new component you made to `StepButton`

SyntaxColoring replied:
Thanks both. I’ll come up with something better for the parent.

> Well actually, it looks like Design System wants the helix component to be called `StepButton` so maybe leave PD's component as `StepContainer` and rename the new component you made to `StepButton`

There’s a comment from Mel in the concurrent module actions Figma file indicating that she plans on renaming it to `StepContainer`, so I’m keeping it there for now.

SyntaxColoring replied:
Renamed to `ConnectedStepContainer` for now. I'd like to come back to this later when I understand more about `ConnectedStepInfo` and how this differs from it.

SyntaxColoring commented on protocol-designer/src/pages/Designer/ProtocolSteps/Timeline/StepContainer.tsx:
```diff
@@ -249,70 +221,35 @@ export function StepContainer(props: StepContainerProps): JSX.Element {
-            ) : null}
-          </Flex>
-        </Box>
+            handleOverflowClick(e)
```
I also struggled with figuring out where to draw the line between what belongs in `components/` and what belongs in `protocol-designer/`. As written:

* The overflow menu *button* is added by `components/`, but when you click that, the overflow menu *itself* is added by `protocol-designer/`.
* The drag-and-drop accoutrements are added by `protocol-designer/`.

Does this seem right?

mjhuff replied:
To each point:

- This seems strange from an engineering perspective, agreed, but maybe this makes sense from a Design perspective (the ones who really own componentization). If one had the bandwidth and want to sync with them over this, it seems valid.
- Yeah this seems ok.

SyntaxColoring commented on components/src/molecules/StepContainer/index.tsx:
I'm calling this a molecule, as opposed to an atom, based on vibes and based on it containing an overflow button.

1. Is molecule right?
2. Is there a better way to answer this question? Like, does Figma indicate atom/molecule/organism anywhere?

SyntaxColoring replied:
Also, is there a better place to put this that indicates it's a thing specific to Protocol Designer?

mjhuff replied:
> Questions 1 & 2

Yes, it's a molecule. I _think_ there's a confluence doc on this somewhere, but I can't find it. My guess is @koji would know where it is.

> Question 3

I don't think so, no.

SyntaxColoring commented on components/src/molecules/StepContainer/index.tsx:
```diff
@@ -0,0 +1,147 @@
+  )
+}
+
+const ELLIPSIZE_STYLE = css`
+  white-space: nowrap;
+  overflow: hidden;
+  text-overflow: ellipsis;
+`
```
Something about the mere presence of styled-components here is causing a million test failures. Anyone know what the deal is?

I'm happy to port the whole component to CSS Modules if we want, but I was hoping to only do one thing at once. Also, plenty of other things in `components/` use styled-components just like this, so I'm not sure why this specifically is breaking things.

SyntaxColoring replied:
Looks like this was a weird circular dependency issue. Styled-components was just the victim. Thanks @mjhuff for the debugging help.

SyntaxColoring committed:
Resolve circular dependency.

SyntaxColoring committed:
Fix missing data-testid.

mjhuff reviewed (approved):
All makes sense to me. See replies to Qs. Thanks!

jerader commented on components/src/molecules/StepContainer/index.tsx:
```diff
@@ -0,0 +1,145 @@
+import { SPACING } from '../../ui-style-constants'
+
+import type { MouseEvent as ReactMouseEvent } from 'react'
+import type { CURSOR_DEFAULT, CURSOR_POINTER, IconName } from '../..'
```
nit: can you import these from their full import? like `../../cursor` and wherever IconName comes from? It can help avoid circular dependency issues down the line since a lot of components can be imported from `../..`

jerader commented on components/src/molecules/StepContainer/StepContainer.stories.tsx:
```diff
@@ -0,0 +1,25 @@
+export default meta
+type Story = StoryObj<typeof meta>
+
+export const StepContainer: Story = {
```
this works but would be great to also make a story for the active, error, and hover states! so design can see each state

SyntaxColoring replied:
The properties are individually configurable in the Storybook UI. Do you still want a set of preconfigured stories, besides that?

<img width="1370" height="801" alt="Screenshot 2025-08-15 at 1 56 55 PM" src="https://github.com/user-attachments/assets/c329e689-09ea-4729-8b78-c45f6143f3c0" />

jerader reviewed (commented):
you did it! i kept putting this task off because i thought it would take so much work. looks great visually in PD and storybook. 

Just a couple of questions to be answered/addressed and this is good to go! @koji when you have time, can you review as well?

SyntaxColoring committed:
Import icon and cursor types from their underlying files instead of index.ts.

SyntaxColoring committed:
Tweak unit test to avoid mocking out the inner OverflowBtn.

Something was going technically wrong with the partial mock-out of @opentrons/components -- it was not replacing OverflowBtn with our mock. I have no idea why or if that would have been solvable. The solution here is to let it render an actual OverflowBtn and look for it via data-testid.

SyntaxColoring committed:
Rename protocol-designer/.../StepContainer -> .../ConnectedStepContainer
